### PR TITLE
Add access log handler for HTTP stream filter.

### PR DESCRIPTION
**Pull request by qiwzhang**
For https://github.com/lyft/envoy/issues/393

### CODE REVIEW

**Comment by mattklein123** [comment]
this looks reasonable to me. tests please.

File: include/envoy/http/filter.h [hunk 1]
```diff
@@ -268,6 +268,22 @@ class StreamFilter : public StreamDecoderFilter, public StreamEncoderFilter {};
...
+ */
+class AccessLogHandler {
+public:
+  virtual ~AccessLogHandler() {}
+
+  /**
+   * Called with request_info data.
+   * @param request_info: the request_info data.
+   */
+  virtual void logRequestInfo(const Http::AccessLog::RequestInfo& request_info) PURE;
```
Comment (mattklein123) [inline include/envoy/http/filter.h]: can we also pass request/response headers (optionally) like the other access log calls? Might be useful for others in the future.

**Comment by qiwzhang** [comment]
@RomanDzhabarov and @mattklein123 
could you help to review it?

**Comment by qiwzhang** [comment]
@RomanDzhabarov and @mattklein123.  I changed to use Http::Accesslog::Instance directly and added a test case.  Could you help to review it again?  Thanks.

File: include/envoy/http/filter.h [hunk 2]
```diff
@@ -292,6 +292,12 @@ class FilterChainFactoryCallbacks {
    * @param filter supplies the filter to add.
    */
   virtual void addStreamFilter(Http::StreamFilterPtr filter) PURE;
+
+  /**
+   * Add an access log handler that is called after response is send.
```
Comment (mattklein123) [inline include/envoy/http/filter.h]: "Add an access log handler that is called when the stream is destroyed." (A response may not have been sent)

Comment (qiwzhang) [inline include/envoy/http/filter.h]: Done

**Pull request merged**